File: packages/sdk-python/agentlens/patchers/openai.py
"""OpenAI auto-patcher for AgentLens.

Importing this module monkey-patches the OpenAI SDK so that every
``chat.completions.create`` (sync and async, streaming and non-streaming)
and ``embeddings.create`` call is automatically traced.

Usage::

    import agentlens.patchers.openai   # patches globally on import

Requires the ``openai`` extra::

    pip install agentlens[openai]
"""

from __future__ import annotations

import uuid
from typing import Any

from ..agentlens import AgentLens
from ..context import get_current_span_id, get_current_trace_id
from ..span import Span

# ---------------------------------------------------------------------------
# Pricing table (mirrors sdk-openai/src/pricing.ts)
# ---------------------------------------------------------------------------

_PRICING: dict[str, tuple[float, float]] = {
    # (input_cost_per_1k, output_cost_per_1k)
    "gpt-4o": (0.005, 0.015),
    "gpt-4o-2024-05-13": (0.005, 0.015),
    "gpt-4o-mini": (0.00015, 0.0006),
    "gpt-4o-mini-2024-07-18": (0.00015, 0.0006),
    "gpt-4-turbo": (0.01, 0.03),
    "gpt-4-turbo-2024-04-09": (0.01, 0.03),
    "gpt-4-turbo-preview": (0.01, 0.03),
    "gpt-4": (0.03, 0.06),
    "gpt-4-0613": (0.03, 0.06),
    "gpt-4-32k": (0.06, 0.12),
    "gpt-3.5-turbo": (0.0005, 0.0015),
    "gpt-3.5-turbo-0125": (0.0005, 0.0015),
    "gpt-3.5-turbo-instruct": (0.0015, 0.002),
    "text-embedding-3-small": (0.00002, 0.0),
    "text-embedding-3-large": (0.00013, 0.0),
    "text-embedding-ada-002": (0.0001, 0.0),
}


def _calculate_cost(model: str, input_tokens: int, output_tokens: int) -> float | None:
    import re

    pricing = _PRICING.get(model) or _PRICING.get(
        re.sub(r"-\d{4}-\d{2}-\d{2}$", "", model)
    )
    if pricing is None:
        return None
    return (input_tokens / 1000) * pricing[0] + (output_tokens / 1000) * pricing[1]


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _make_span(name: str) -> Span | None:
    """Create a raw Span using the current trace context."""
    project_id = AgentLens._get_project_id()
    if not project_id:
        return None
    trace_id = get_current_trace_id() or str(uuid.uuid4())
    parent_span_id = get_current_span_id()
    return Span(str(uuid.uuid4()), trace_id, name, project_id, parent_span_id)


def _finish_span(span: Span) -> None:
    span.end()
    AgentLens._push_span(span.to_dict())


# ---------------------------------------------------------------------------
# Streaming wrappers
# ---------------------------------------------------------------------------

def _wrap_stream_sync(span: Span, stream: Any, model: str) -> Any:
    """Wrap a sync streaming response, accumulating content for the span."""
    content = ""
    prompt_tokens = 0
    completion_tokens = 0
    try:
        for chunk in stream:
            delta = ""
            if chunk.choices:
                delta = chunk.choices[0].delta.content or ""
            content += delta
            if hasattr(chunk, "usage") and chunk.usage:
                prompt_tokens = chunk.usage.prompt_tokens or 0
                completion_tokens = chunk.usage.completion_tokens or 0
            yield chunk
        if content:
            span.set_output(content)
        if prompt_tokens or completion_tokens:
            cost = _calculate_cost(model, prompt_tokens, completion_tokens)
            span.set_tokens(prompt_tokens, completion_tokens, cost)
    except Exception as exc:
        span.set_error(exc)
        raise
    finally:
        _finish_span(span)


async def _wrap_stream_async(span: Span, stream: Any, model: str) -> Any:
    """Wrap an async streaming response, accumulating content for the span."""
    content = ""
    prompt_tokens = 0
    completion_tokens = 0
    try:
        async for chunk in stream:
            delta = ""
            if chunk.choices:
                delta = chunk.choices[0].delta.content or ""
            content += delta
            if hasattr(chunk, "usage") and chunk.usage:
                prompt_tokens = chunk.usage.prompt_tokens or 0
                completion_tokens = chunk.usage.completion_tokens or 0
            yield chunk
        if content:
            span.set_output(content)
        if prompt_tokens or completion_tokens:
            cost = _calculate_cost(model, prompt_tokens, completion_tokens)
            span.set_tokens(prompt_tokens, completion_tokens, cost)
    except Exception as exc:
        span.set_error(exc)
        raise
    finally:
        _finish_span(span)


# ---------------------------------------------------------------------------
# Patch bookkeeping
# ---------------------------------------------------------------------------

_patches: list[tuple[Any, str, Any]] = []  # (obj, attr_name, original)
_patched = False


def _save_patch(obj: Any, attr: str, replacement: Any) -> None:
    original = getattr(obj, attr)
    _patches.append((obj, attr, original))
    setattr(obj, attr, replacement)


# ---------------------------------------------------------------------------
# chat.completions.create
# ---------------------------------------------------------------------------

def _patch_chat_completions(completions_cls: Any) -> None:
    original_create = completions_cls.create

    def patched_create(self: Any, *args: Any, **kwargs: Any) -> Any:
        if not AgentLens._is_initialized():
            return original_create(self, *args, **kwargs)
        span = _make_span("openai.chat.completions")
        if span is None:
            return original_create(self, *args, **kwargs)

        model: str = kwargs.get("model", args[0] if args else "")
        messages = kwargs.get("messages")
        if messages:
            import json as _json
            span.set_model(model, "openai")
            span.set_input(_json.dumps(messages))

        stream = kwargs.get("stream", False)

        if stream:
            try:
                raw = original_create(self, *args, **kwargs)
            except Exception as exc:
                span.set_error(exc)
                _finish_span(span)
                raise
            return _wrap_stream_sync(span, raw, model)

        try:
            result = original_create(self, *args, **kwargs)
            if result.choices:
                content = result.choices[0].message.content or ""
                if content:
                    span.set_output(content)
            if result.usage:
                cost = _calculate_cost(
                    model, result.usage.prompt_tokens, result.usage.completion_tokens
                )
                span.set_tokens(result.usage.prompt_tokens, result.usage.completion_tokens, cost)
            return result
        except Exception as exc:
            span.set_error(exc)
            raise
        finally:
            _finish_span(span)

    _save_patch(completions_cls, "create", patched_create)


def _patch_async_chat_completions(completions_cls: Any) -> None:
    original_create = completions_cls.create

    async def patched_create(self: Any, *args: Any, **kwargs: Any) -> Any:
        if not AgentLens._is_initialized():
            return await original_create(self, *args, **kwargs)
        span = _make_span("openai.chat.completions")
        if span is None:
            return await original_create(self, *args, **kwargs)

        model: str = kwargs.get("model", args[0] if args else "")
        messages = kwargs.get("messages")
        if messages:
            import json as _json
            span.set_model(model, "openai")
            span.set_input(_json.dumps(messages))

        stream = kwargs.get("stream", False)

        if stream:
            try:
                raw = await original_create(self, *args, **kwargs)
            except Exception as exc:
                span.set_error(exc)
                _finish_span(span)
                raise
            return _wrap_stream_async(span, raw, model)

        try:
            result = await original_create(self, *args, **kwargs)
            if result.choices:
                content = result.choices[0].message.content or ""
                if content:
                    span.set_output(content)
            if result.usage:
                cost = _calculate_cost(
                    model, result.usage.prompt_tokens, result.usage.completion_tokens
                )
                span.set_tokens(result.usage.prompt_tokens, result.usage.completion_tokens, cost)
            return result
        except Exception as exc:
            span.set_error(exc)
            raise
        finally:
            _finish_span(span)

    _save_patch(completions_cls, "create", patched_create)


# ---------------------------------------------------------------------------
# embeddings.create
# ---------------------------------------------------------------------------

def _patch_embeddings(embeddings_cls: Any) -> None:
    original_create = embeddings_cls.create

    def patched_create(self: Any, *args: Any, **kwargs: Any) -> Any:
        if not AgentLens._is_initialized():
            return original_create(self, *args, **kwargs)
        span = _make_span("openai.embeddings")
        if span is None:
            return original_create(self, *args, **kwargs)

        model: str = kwargs.get("model", args[0] if args else "")
        span.set_model(model, "openai")
        inp = kwargs.get("input")
        if isinstance(inp, str):
            span.set_input(inp)
        elif isinstance(inp, list):
            import json as _json
            span.set_input(_json.dumps(inp))

        try:
            result = original_create(self, *args, **kwargs)
            if result.usage:
                cost = _calculate_cost(model, result.usage.prompt_tokens, 0)
                span.set_tokens(result.usage.prompt_tokens, 0, cost)
            return result
        except Exception as exc:
            span.set_error(exc)
            raise
        finally:
            _finish_span(span)

    _save_patch(embeddings_cls, "create", patched_create)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def patch() -> None:
    """Monkey-patch the OpenAI SDK to auto-trace all LLM calls.

    Safe to call multiple times — subsequent calls are no-ops.
    Called automatically when this module is imported.
    """
    global _patched
    if _patched:
        return

    try:
        import openai as _openai
    except ImportError:
        import warnings
        warnings.warn(
            "[AgentLens] openai package not found — skipping auto-patch.",
            stacklevel=2,
        )
        return

    # Sync client
    probe = _openai.OpenAI(api_key="__agentlens_probe__")
    _patch_chat_completions(type(probe.chat.completions))
    _patch_embeddings(type(probe.embeddings))

    # Async client
    async_probe = _openai.AsyncOpenAI(api_key="__agentlens_probe__")
    _patch_async_chat_completions(type(async_probe.chat.completions))

    _patched = True


def unpatch() -> None:
    """Restore all patched OpenAI methods to their original implementations.

    Primarily used in tests to ensure a clean state between test cases.
    """
    global _patched
    for obj, attr, original in _patches:
        setattr(obj, attr, original)
    _patches.clear()
    _patched = False


# Auto-patch on import
patch()

File: packages/sdk-universal/src/parsers/openai.ts
import type { ParsedSpan } from '../types'

const OPENAI_COSTS: Record<string, { input: number; output: number }> = {
  'gpt-4o': { input: 0.0000025, output: 0.000010 },
  'gpt-4o-mini': { input: 0.00000015, output: 0.0000006 },
  'gpt-4-turbo': { input: 0.000010, output: 0.000030 },
  'gpt-4': { input: 0.000030, output: 0.000060 },
  'gpt-3.5-turbo': { input: 0.0000005, output: 0.0000015 },
  'text-embedding-3-small': { input: 0.00000002, output: 0 },
  'text-embedding-3-large': { input: 0.00000013, output: 0 },
}

function lookupCost(model: string): { input: number; output: number } {
  if (OPENAI_COSTS[model]) return OPENAI_COSTS[model]
  // Permit version-suffixed model names like "gpt-4o-2024-08-06"
  for (const key of Object.keys(OPENAI_COSTS)) {
    if (model.startsWith(key)) return OPENAI_COSTS[key]
  }
  return { input: 0, output: 0 }
}

function extractInputText(request: Record<string, unknown> | null): string {
  if (!request) return ''
  const messages = request.messages
  if (Array.isArray(messages)) {
    return messages
      .map((m) => {
        const content = (m as { content?: unknown }).content
        if (typeof content === 'string') return content
        if (Array.isArray(content)) {
          return content
            .map((part) => (typeof (part as { text?: unknown }).text === 'string' ? (part as { text: string }).text : ''))
            .join('')
        }
        return ''
      })
      .join('\n')
  }
  if (typeof request.prompt === 'string') return request.prompt
  if (Array.isArray(request.input)) return (request.input as unknown[]).map((x) => String(x)).join('\n')
  if (typeof request.input === 'string') return request.input
  return ''
}

function extractOutputText(response: Record<string, unknown> | null): string {
  if (!response) return ''
  const choices = response.choices
  if (Array.isArray(choices) && choices.length > 0) {
    const first = choices[0] as { message?: { content?: string }; text?: string }
    if (first.message?.content) return first.message.content
    if (typeof first.text === 'string') return first.text
  }
  // embeddings: no text output
  return ''
}

export function parseOpenAI(args: {
  request: Record<string, unknown> | null
  response: Record<string, unknown> | null
  isStream: boolean
}): ParsedSpan {
  const { request, response, isStream } = args

  const model = (response?.model as string) ?? (request?.model as string) ?? 'unknown'
  const usage = (response?.usage ?? {}) as {
    prompt_tokens?: number
    completion_tokens?: number
    total_tokens?: number
  }

  const inputTokens = usage.prompt_tokens ?? 0
  const outputTokens = usage.completion_tokens ?? 0
  const totalTokens = usage.total_tokens ?? inputTokens + outputTokens

  const cost = lookupCost(model)
  const costUsd = inputTokens * cost.input + outputTokens * cost.output

  return {
    model,
    provider: 'openai',
    inputTokens,
    outputTokens,
    totalTokens,
    costUsd,
    inputText: extractInputText(request),
    outputText: extractOutputText(response),
    isStream,
  }
}

export function parseOpenAIStream(args: {
  request: Record<string, unknown> | null
  outputText: string
  usage: { prompt_tokens?: number; completion_tokens?: number; total_tokens?: number }
}): ParsedSpan {
  const { request, outputText, usage } = args
  const model = (request?.model as string) ?? 'unknown'
  const inputTokens = usage.prompt_tokens ?? 0
  const outputTokens = usage.completion_tokens ?? 0
  const totalTokens = usage.total_tokens ?? inputTokens + outputTokens
  const cost = lookupCost(model)

  return {
    model,
    provider: 'openai',
    inputTokens,
    outputTokens,
    totalTokens,
    costUsd: inputTokens * cost.input + outputTokens * cost.output,
    inputText: extractInputText(request),
    outputText,
    isStream: true,
  }
}
